src/app/services/baw-api/base-api.service.ts

```diff
@@ -35,12 +35,56 @@ export abstract class BawApiService {
     filter -> POST with filter body
   */
 
-  private url = this.config.getConfig().environment.apiRoot;
-  protected userSessionStorage = "user";
-  protected id = (x: number) => x;
-  protected param = (x: string) => x;
+  protected userSessionStorage = "baw.client.user";
 
-  constructor(private http: HttpClient, private config: AppConfigService) {}
+  /**
+   * Handle API response
+   */
+  private handleResponse: (_: ApiResponse<T>) => T[];
+  private handleSingleResponse: (_: ApiResponse<T>) => T;
+
+  // * @param nullResponse True if response body can be empty. If set, next always be called with "true"
+  // {
+
+  // next: <T>(response: ApiResponse<T>) => {
+  //   if (response?.data) {
+  //     next(response.data);
+  //   } else if (!response && nullResponse) {
+  //     // TODO Remove if https://github.com/QutEcoacoustics/baw-server/issues/427 is fixed
+  //     next(true);
+  //   } else {
+  //     error({
+  //       status: apiReturnCodes.unknown,
+  //       message: "No data returned from API"
+  //     } as ApiErrorDetails);
+  //   }
+  // },
+  // error: (err: ApiErrorDetails) => {
+  //   error(err);
+  // }
+  // };
+
+  constructor(
+    protected http: HttpClient,
+    private apiRoot: string,
+    _new: new (_: object) => T) {
+
+    this.handleSingleResponse = (response: ApiResponse<T>) => {
+      if (response.data instanceof Array) {
+        throw new Error("Received an array of API results when only a single result was expected");
+      }
+
+      return new _new(response.data);
+    };
+
+    this.handleResponse = (response: ApiResponse<T>): T[] => {
+      if (response.data instanceof Array) {
+        return response.data.map(x => new _new(x));
+      } else {
+        return [new _new(response.data)];
+      }
+    };
```

Comment:
Not sure why these are implemented in the constructor

Reply:
Ah, that was a tricky bug. 

Reactive requires pure functions. Just pulling a function off of the service class (i.e. `this.handleResponse`) risks rebinding `this`. We need `this` though so we can get access to that `_new` property.

By assigning arrow functions to the values in the constructor, we close over `_new` instead and do not need to rely on `this`.